pyteal/ast/gitxn_test.py

```diff
@@ -21,10 +21,21 @@ def test_gitxn_invalid():
 
 
 def test_gitxn_valid():
-    GitxnaExpr(0, TxnField.application_args, 1).__teal__(teal6Options)
-
     for i in range(MAX_GROUP_SIZE):
         Gitxn[i].sender()
 
 
+def test_gitxna_invalid():
+    with pytest.raises(TealInputError):
+        GitxnaExpr("Invalid_type", TxnField.application_args, 1).__teal__(teal6Options)
+
+    with pytest.raises(TealInputError):
+        GitxnaExpr(0, TxnField.application_args, "Invalid_type").__teal__(teal6Options)
+
+
+def test_gitxna_valid():
+    GitxnaExpr(0, TxnField.application_args, 1).__teal__(teal6Options)
+    GitxnaExpr(0, TxnField.application_args, Assert(Int(1))).__teal__(teal6Options)
```

Comment:
This exposes an invalid case we aren't currently checking for any txna opcodes currently. It _should_ be invalid to use an `Expr` that evaluates to anything but a `TealType.uint64` for the index here. `Assert(Int(1))` returns an Expr with the type `TealType.none`, so we should error.

Ideally `TxnaExpr.__init__` would check this with `require_type(index, TealType.uint64)` when `index` is not a Python int. And I suppose the same problem exists for `GtxnExpr` and `GtxnaExpr` with respect to `txnIndex`, so they should have a similar check as well.

(The reason for erroring in the `__init__` method is so that a reasonable stack trace is generated which will likely contain code that the user explicitly called to create their program. If we fail later, the stack trace will come from the `compileTeal` method, which is less useful.)

We could also modify `pyteal/ast/txn_test.py` to make sure an error is raised if a non-`TealType.uint64` expression is given as an index.

@michaeldiamant if you feel confident enough to make those changes in this PR, feel free to do so. If not, it can be addressed in a different PR.

Reply:
@jasonpaulos Feedback makes sense to me.  I had wondered _why_ `Expr` was applied so loosely.  

Since the changes felt large enough to me, I opened a child PR to isolate my attempt to address the feedback:  https://github.com/michaeldiamant/pyteal/pull/1. 

If you feel the changes to test setup are counterproductive, let me know and I'll back out stylistic changes.

Reply:
Resolving because the child PR (https://github.com/michaeldiamant/pyteal/pull/1) is merged into this PR.